MachObfuscator/Controller/Obfuscator.swift

```diff
@@ -1,28 +1,42 @@
 import Foundation
 
 class Obfuscator {
-    private let directoryURL: URL
+    private let directoryOrFileURL: URL
+    private let isDirectory: Bool
 
     private let mangler: SymbolMangling
 
     private let options: Options
 
-    init(directoryURL: URL, mangler: SymbolMangling, options: Options) {
-        self.directoryURL = directoryURL
+    init(directoryOrFileURL: URL, mangler: SymbolMangling, options: Options) {
+        var isDir: ObjCBool = false
+        guard FileManager.default.fileExists(atPath: directoryOrFileURL.path, isDirectory: &isDir) else {
+            fatalError("'\(directoryOrFileURL.path)' does not exist")
+        }
+        self.directoryOrFileURL = directoryOrFileURL
+        isDirectory = isDir.boolValue
+
         self.mangler = mangler
         self.options = options
     }
 
     func run(loader: ImageLoader & SymbolsSourceLoader & DependencyNodeLoader = SimpleImageLoader(),
              sourceSymbolsLoader: SourceSymbolsLoader = SimpleSourceSymbolsLoader()) {
-        LOGGER.info("Will obfuscate \(directoryURL)")
+        LOGGER.info("Will obfuscate \(directoryOrFileURL)")
 
         LOGGER.info("Looking for dependencies...")
-        let paths = time(withTag: "Looking for dependencies") {
-            ObfuscationPaths.forAllExecutables(inDirectory: directoryURL,
-                                               dependencyNodeLoader: loader,
-                                               obfuscableFilesFilter: options.obfuscableFilesFilter,
-                                               withDependencies: options.analyzeDependencies)
+        let paths = time(withTag: "Looking for dependencies") { () -> ObfuscationPaths in
+            if isDirectory {
+                return ObfuscationPaths.forAllExecutables(inDirectory: directoryOrFileURL,
+                                                          dependencyNodeLoader: loader,
+                                                          obfuscableFilesFilter: options.obfuscableFilesFilter,
+                                                          withDependencies: options.analyzeDependencies)
+            } else {
+                return ObfuscationPaths.forExecutable(executable: directoryOrFileURL,
```

Comment:
The documentation is about the file, the logic is about the file, but the method parameter says about executable. Is that correct? Shouldn't this parameter be named e.g. `machoFile`?

Reply:
method and parameters names were supposed to be consistent with existing methods in `ObfuscationPaths+Building.swift` - that is why it is named `executable`.
`forExecutable` accepts any mach-o file, not only the ones of type `MH_EXECUTE` and in theory it could obfuscate a dylib, so the name may be a bit misleading, but the same is true for `addExecutable` method.

If you like, I may change `forExecutable(executable:)` to `forMachOFile(machOFile:)`

Reply:
> forExecutable accepts any mach-o file, not only the ones of type MH_EXECUTE and in theory it could obfuscate a dylib
Yes, it will accept any mach-o, but it was meant to accept just executables – it is a private and was called just by the `addAllExecutables`.

After taking a look at `addExecutable` I can see this is more tricky case. Passed url is used as an `@executable_path` rpath token. Assuming that the user wants to obfuscate a mach-o file anyway, MachObfuscator could at least emit a warning in that case. What do you think?

So I think, that we should either:
- rename `forExecutable(executable:)` to `for(machOFile machOFileURL:)` (I think we should not repeat first parameter in function name),
- change `RpathsAccumulator.executablePath` to be optional
- make `RpathsAccumulator.executablePath` nil when the root image is not an executable
- crash when `RpathsAccumulator.executablePath` is nil and any dependent image uses `@executable_path` rpath token.

OR:

- add requirement for `appDirectoryOrFile` to be either a directory or a executable file,
- change the documentation to express that requirement.

What do you think?

Reply:
I cannot see why do we need an exception for `@executable_path` when argument passed to Obfuscator points directly to executable file. In `addExecutable` the directory is calculated relative to path of currently processed executable (`let executableDir = executableURL.deletingLastPathComponent()`), regardless if it is part of extracted IPA or standalone executable. As I understand, this is correct path.

If user obfuscates single file and this file happen to load some libs relative to `@executable_path` (eg. `../Frameworks/sth`), the libs will be found and treated as system dependencies which is correct for the needs of obfuscation. I think it is a desired behaviour.

> add requirement for `appDirectoryOrFile` to be either a directory or a executable file

`appDirectoryOrFile` should a directory or an executable file, otherwise nothing will be obfuscated.

Reply:
`for` is a bad method name, because it is a keyword, I will leave `forExecutable` and change only the argument name